### PR TITLE
TDR-2616 - Update terraform for new backend checks architecture

### PR DESCRIPTION
**Pull request by vimleshtna**
- Create new lambda functions for the existing checks
- Create new security groups
- Create a step function
- Update the GitHub actions lambda deploy role for the new lambda functions.

### CODE REVIEW

File: root_backend_checks.tf [hunk 1]
```diff
@@ -24,3 +24,247 @@ module "backend_checks_api" {
...
+  ]
+}
+
+module "yara_av_v2" {
+  source        = "./tdr-terraform-modules/generic_lambda"
+  tags          = local.common_tags
+  function_name = local.yara_av_v2_function_name
+  handler       = "matcher.matcher_lambda_handler"
+  policies = {
+    "TDRChecksumV2LambdaPolicy${title(local.environment)}" = templatefile("./templates/iam_policy/lambda_s3_only_policy.json.tpl", {
```
Comment (TomJKing) [inline root_backend_checks.tf:223]: Is that the correct policy name, should it not be `TDRYaraAVV2LambdaPolicy...`?

File: templates/iam_policy/allow_iam_db_auth.json.tpl [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+      ],
+      "Resource": [
+        "arn:aws:logs:eu-west-2:${account_id}:log-group:/aws/lambda/${function_name}",
+        "arn:aws:logs:eu-west-2:${account_id}:log-group:/aws/lambda/${function_name}:log-stream:*"
+      ]
+    },
+    {
+      "Effect": "Allow",
+      "Action": [
+        "ec2:CreateNetworkInterface",
```
Comment (TomJKing) [inline templates/iam_policy/allow_iam_db_auth.json.tpl:18]: Check alphabetical ordering of the `Action`s

File: templates/iam_policy/backend_check_policy.json.tpl [hunk 3]
```diff
@@ -0,0 +1,21 @@
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Action": [
+        "lambda:InvokeFunction"
+      ],
+      "Resource": [
+        "${file_format_v2_arn}",
```
Comment (TomJKing) [inline templates/iam_policy/backend_check_policy.json.tpl:10]: Could you order these alphabetically by parameter name.

File: templates/iam_policy/deploy_lambda_github_actions.json.tpl [hunk 4]
```diff
@@ -33,6 +33,13 @@
         "arn:aws:lambda:${region}:${account_id}:function:tdr-signed-cookies-${environment}",
         "arn:aws:lambda:${region}:${account_id}:function:tdr-yara-av-${environment}",
         "arn:aws:lambda:${region}:${account_id}:function:tdr-reporting-${environment}",
+        "arn:aws:lambda:${region}:${account_id}:function:tdr-file-upload-data-${environment}",
```
Comment (TomJKing) [inline templates/iam_policy/deploy_lambda_github_actions.json.tpl:36]: Could you put these in alphabetical order with the existing lambda function names.

File: templates/iam_policy/lambda_policy.json.tpl [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+      ],
+      "Resource": [
+        "arn:aws:logs:eu-west-2:${account_id}:log-group:/aws/lambda/${function_name}",
+        "arn:aws:logs:eu-west-2:${account_id}:log-group:/aws/lambda/${function_name}:log-stream:*"
+      ]
+    },
+    {
+      "Effect": "Allow",
+      "Action": [
+        "ec2:CreateNetworkInterface",
```
Comment (TomJKing) [inline templates/iam_policy/lambda_policy.json.tpl:18]: Check alphabetical ordering of the `Action`s

File: templates/iam_policy/lambda_s3_only_policy.json.tpl [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+      ],
+      "Resource": [
+        "arn:aws:logs:eu-west-2:${account_id}:log-group:/aws/lambda/${function_name}",
+        "arn:aws:logs:eu-west-2:${account_id}:log-group:/aws/lambda/${function_name}:log-stream:*"
+      ]
+    },
+    {
+      "Effect": "Allow",
+      "Action": [
+        "ec2:CreateNetworkInterface",
```
Comment (TomJKing) [inline templates/iam_policy/lambda_s3_only_policy.json.tpl:18]: Check alphabetical ordering of the `Action`s

File: templates/iam_policy/lambda_s3_policy.json.tpl [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+      ],
+      "Resource": [
+        "arn:aws:logs:eu-west-2:${account_id}:log-group:/aws/lambda/${function_name}",
+        "arn:aws:logs:eu-west-2:${account_id}:log-group:/aws/lambda/${function_name}:log-stream:*"
+      ]
+    },
+    {
+      "Effect": "Allow",
+      "Action": [
+        "ec2:CreateNetworkInterface",
```
Comment (TomJKing) [inline templates/iam_policy/lambda_s3_policy.json.tpl:18]: Check alphabetical ordering of the `Action`s

File: templates/step_function/backend_checks_definition.json.tpl [hunk 8]
```diff
@@ -0,0 +1,279 @@
+{
+  "Comment": "A description of my state machine",
```
Comment (TomJKing) [inline templates/step_function/backend_checks_definition.json.tpl:2]: Could the `Comment` be updated to describe what the state machine is

File: templates/step_function/backend_checks_definition.json.tpl [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+      "Resource": "arn:aws:states:::lambda:invoke",
+      "OutputPath": "$.Payload",
+      "Parameters": {
+        "Payload.$": "$",
+        "FunctionName": "${file_upload_data_arn}"
+      },
+      "Retry": [
+        {
+          "ErrorEquals": [
+            "Lambda.ServiceException",
```
Comment (TomJKing) [inline templates/step_function/backend_checks_definition.json.tpl:16]: Could you order these alphabetically.

File: templates/step_function/backend_checks_definition.json.tpl [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+          "redactedResults": {
+            "redactedFiles": [],
+            "errors": []
+          }
+        }
+      },
+      "Retry": [
+        {
+          "ErrorEquals": [
+            "Lambda.ServiceException",
```
Comment (TomJKing) [inline templates/step_function/backend_checks_definition.json.tpl:44]: Could you order these alphabetically.

File: templates/step_function/backend_checks_definition.json.tpl [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+                    "Resource": "arn:aws:states:::lambda:invoke",
+                    "OutputPath": "$.Payload",
+                    "Parameters": {
+                      "FunctionName": "${yara_av_v2_arn}",
+                      "Payload.$": "$"
+                    },
+                    "Retry": [
+                      {
+                        "ErrorEquals": [
+                          "Lambda.ServiceException",
```
Comment (TomJKing) [inline templates/step_function/backend_checks_definition.json.tpl:79]: Could you order these alphabetically

File: templates/step_function/backend_checks_definition.json.tpl [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+                    "Resource": "arn:aws:states:::lambda:invoke",
+                    "OutputPath": "$.Payload",
+                    "Parameters": {
+                      "FunctionName": "${file_format_v2_arn}",
+                      "Payload.$": "$"
+                    },
+                    "Retry": [
+                      {
+                        "ErrorEquals": [
+                          "Lambda.ServiceException",
```
Comment (TomJKing) [inline templates/step_function/backend_checks_definition.json.tpl:106]: Could you order these alphabetically.

File: templates/step_function/backend_checks_definition.json.tpl [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+                    "Resource": "arn:aws:states:::lambda:invoke",
+                    "OutputPath": "$.Payload",
+                    "Parameters": {
+                      "FunctionName": "${checksum_v2_arn}",
+                      "Payload.$": "$"
+                    },
+                    "Retry": [
+                      {
+                        "ErrorEquals": [
+                          "Lambda.ServiceException",
```
Comment (TomJKing) [inline templates/step_function/backend_checks_definition.json.tpl:133]: Could you order these alphabetically.

File: templates/step_function/backend_checks_definition.json.tpl [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+      "Type": "Task",
+      "Resource": "arn:aws:states:::lambda:invoke",
+      "Parameters": {
+        "Payload.$": "$",
+        "FunctionName": "${redacted_files_arn}"
+      },
+      "Retry": [
+        {
+          "ErrorEquals": [
+            "Lambda.ServiceException",
```
Comment (TomJKing) [inline templates/step_function/backend_checks_definition.json.tpl:182]: Could you order these alphabetically.

File: templates/step_function/backend_checks_definition.json.tpl [hunk 15]
```diff
@@ -0,0 +1,279 @@
...
+      "Type": "Task",
+      "Resource": "arn:aws:states:::lambda:invoke",
+      "Parameters": {
+        "Payload.$": "$",
+        "FunctionName": "${statuses_arn}"
+      },
+      "Retry": [
+        {
+          "ErrorEquals": [
+            "Lambda.ServiceException",
```
Comment (TomJKing) [inline templates/step_function/backend_checks_definition.json.tpl:209]: Could you order these alphabetically.

File: templates/step_function/backend_checks_definition.json.tpl [hunk 16]
```diff
@@ -0,0 +1,279 @@
...
+      "Resource": "arn:aws:states:::lambda:invoke",
+      "OutputPath": "$.Payload",
+      "Parameters": {
+        "Payload.$": "$",
+        "FunctionName": "${api_update_v2_arn}"
+      },
+      "Retry": [
+        {
+          "ErrorEquals": [
+            "Lambda.ServiceException",
```
Comment (TomJKing) [inline templates/step_function/backend_checks_definition.json.tpl:236]: Could you order these alphabetically.

File: templates/step_function/backend_checks_definition.json.tpl [hunk 17]
```diff
@@ -0,0 +1,279 @@
...
+      "Type": "Task",
+      "Resource": "arn:aws:states:::lambda:invoke",
+      "OutputPath": "$.Payload",
+      "Parameters": {
+        "FunctionName": "${notification_arn}",
+        "Payload": {
+          "consignmentId.$": "$$.Execution.Input.consignmentId",
+          "error.$": "$.Error",
+          "cause.$": "$.Cause",
+          "environment": "intg"
```
Comment (TomJKing) [inline templates/step_function/backend_checks_definition.json.tpl:257]: Should the `environment` field have a hard coded value?

File: templates/step_function/backend_checks_definition.json.tpl [hunk 18]
```diff
@@ -0,0 +1,279 @@
...
+        }
+      },
+      "Retry": [
+        {
+          "ErrorEquals": [
+            "Lambda.ServiceException",
```
Comment (TomJKing) [inline templates/step_function/backend_checks_definition.json.tpl:263]: Could you order these alphabetically.

**Review by TomJKing — COMMENTED** [review]
Looks good. Just minor comments. Also assume that this need some tdr-terraform-modules repo changes to work correctly

File: root_backend_checks.tf (hunk 1 above)
Comment (vimleshtna) [reply root_backend_checks.tf:229]: Good spot. I changed it, thanks.

File: templates/step_function/backend_checks_definition.json.tpl (hunk 17 above)
Comment (vimleshtna) [reply templates/step_function/backend_checks_definition.json.tpl]: Good spot. I have updated it, thanks.

**Review by TomJKing — COMMENTED** [review]
Looks good, but looks like you might have a compilation error: 

```
Error: Unsupported block type
│ 
│   on tdr-terraform-modules/generic_lambda/main.tf line 10, in resource "aws_lambda_function" "lambda_function":
│   10:   ephemeral_storage {
```

**Comment by vimleshtna** [comment]
> Looks good, but looks like you might have a compilation error:
> 
> ```
> Error: Unsupported block type
> │ 
> │   on tdr-terraform-modules/generic_lambda/main.tf line 10, in resource "aws_lambda_function" "lambda_function":
> │   10:   ephemeral_storage {
> ```

Yeah, I have upgraded the provider. It's working now.

**Pull request merged**